src/libraries/System.Security.Cryptography.X509Certificates/tests/RevocationTests/TimeoutTests.cs

```diff
@@ -0,0 +1,201 @@
+
+                chain.ChainPolicy.DisableCertificateDownloads = true;
+
+                Stopwatch watch = Stopwatch.StartNew();
+                Assert.True(chain.Build(endEntityCert));
+                watch.Stop();
+
+                // There should be two network fetches, OCSP/CRL to intermediate to get leaf status,
+                // OCSP/CRL to root to get intermediate statuses. It should take at least 2x the delay
+                // plus other non-network time, so we can at least ensure it took as long as
+                // the delay for each fetch.
+                Assert.True(watch.Elapsed >= delay * 2);
+            }
+        }
+
+        [Theory]
+        [InlineData(PkiOptions.OcspEverywhere)]
+        [InlineData(PkiOptions.CrlEverywhere)]
+        [PlatformSpecific(TestPlatforms.Windows | TestPlatforms.Linux)]
+        public static void RevocationCheckingTimeout(PkiOptions pkiOptions)
+        {
+            CertificateAuthority.BuildPrivatePki(
+                pkiOptions,
+                out RevocationResponder responder,
+                out CertificateAuthority rootAuthority,
+                out CertificateAuthority intermediateAuthority,
+                out X509Certificate2 endEntityCert,
+                nameof(RevocationCheckingTimeout));
+
+            using (responder)
+            using (rootAuthority)
+            using (intermediateAuthority)
+            using (endEntityCert)
+            using (ChainHolder holder = new ChainHolder())
+            using (X509Certificate2 rootCert = rootAuthority.CloneIssuerCert())
+            using (X509Certificate2 intermediateCert = intermediateAuthority.CloneIssuerCert())
+            {
+                TimeSpan delay = TimeSpan.FromSeconds(3);
+
+                X509Chain chain = holder.Chain;
+                responder.RespondDelay = delay;
+                responder.DelayedActions = RevocationResponder.DelayedActionsFlag.All;
+
+                chain.ChainPolicy.UrlRetrievalTimeout = TimeSpan.FromSeconds(1);
+                chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+                chain.ChainPolicy.CustomTrustStore.Add(rootCert);
+                chain.ChainPolicy.ExtraStore.Add(intermediateCert);
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.Online;
+                chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+
+                chain.ChainPolicy.DisableCertificateDownloads = true;
+
+                Assert.False(chain.Build(endEntityCert));
+
+                const X509ChainStatusFlags ExpectedFlags =
+                    X509ChainStatusFlags.RevocationStatusUnknown |
+                    X509ChainStatusFlags.OfflineRevocation;
+
+                X509ChainStatusFlags eeFlags = GetFlags(chain, endEntityCert.Thumbprint);
+                X509ChainStatusFlags intermediateFlags = GetFlags(chain, intermediateCert.Thumbprint);
```

Comment:
Why match by thumbprint? You know the expected index.

Reply:
The intent is less clear to me in a test. "I am looking for flags in for a cert with this thumbprint" reads better to me than "Get the flags at 1", also because for the EE I can never remember if it is at 0 or at Length - 1.